Validate forms, sections and cdes when exporting

alintreznai opened the pull request:
Validate forms, sections and CDEs when exporting registries.

Commit:
Validate forms, sections and cdes when exporting

sztamas commented on rdrf/rdrf/services/io/defs/exporter.py:
```diff
@@ -145,6 +161,8 @@ def _create_section_map(self, section_code, optional=False):
             if optional:
                 return {}
             raise
+        if not optional:
```
Why is the if necessary?
If the section didn't exist but was optional we already returned `{}` and won't get to this point.

alintreznai replied:
Fixed.

sztamas commented on rdrf/rdrf/services/io/defs/exporter.py:
```diff
@@ -112,6 +121,13 @@ def _get_cdes(self, export_type):
+        try:
+            model.clean()
+        except ValidationError as verr:
+            raise ExportException(verr.message)
```
Maybe worth adding `from verr` to keep the `ValidationError`? Might be useful.
Probably should be `str(verr)` as I don't think `verr.message` is always a property of the validation error.

Also, the message should contain the model to give us better error messages.
But even if you do that the Export management command has to be changed, to display more helpful error messages. Currently, if the export fails you don't know why it did so.

alintreznai replied:
Updated to include the model in the exception message and wrap the original exception.

sztamas reviewed (commented):
The approach might work ok and it is helpful as it is if we improved the error messages, so we see why the validation failed.

However, it seems a bit hackish. What is the purpose of the new Export Format VALIDATION?
Is it just so we can run validation of a registry by using the Export Format?
If so, I'm not sure I like it. We can just use format None or YAML and ignore the exported data.
Sort of a `--simulate` or `--dry-run`.

As discussed, I would be more inclined to have another class responsible for validating a registry.
I do see that this approach is way easier to implement though and it is better than the current export, so I think it is fine to keep it as long as we improve the error messages and maybe drop the new Export Format.

Commit:
Small cleanup, better exception handling

alintreznai commented:
Updated the code so that validation is separated from the export and it's configurable via an export parameter.

Commit:
Separate validation code from export

sztamas committed:
Don't return a blank page if the export fails.

If there were any errors during export the user got a blank page in the
browser and had to hit back to get back to the Registries Admin page
showing the error message.

Commit:
Small fix and better error handling

Commit:
Fix issue with exporting multiple registries

sztamas committed:
Cosmetic changes.

sztamas committed:
Add more detail to export exceptions.

sztamas commented on rdrf/rdrf/services/io/defs/exporter.py:
```diff
@@ -54,7 +59,39 @@ class Exporter(object):
+        Validates the CDES, sections and registry forms
+        Raises an ExporterException in case of error
+        """
+        cdes = self._get_cdes(ExportType.REGISTRY_PLUS_CDES)
```
Not sure if this is good enough.

We should validate only the CDEs that will be included in the export, but there might be a mismatch if we don't pass in the export type to validate.

Also, I don't see where `form.complete_form_cdes` are validated.

alintreznai replied:
Updated. The complete_form_cdes are validated when the form is validated as it's a part of the form's clean method.

Commit:
Small improvements
